### ISSUE TITLE
 The quantized ONNX model produced with run_ptq.sh is missing an output.

### ISSUE THREAD

**Issue by xXxTheUserxXx** [amd/RyzenAI-SW#51]
Hi,

I'm having trouble reproducing the post-training quantization (run_ptq.sh) for object detection in the tutorial/yolov8_e2e.

I don't encounter any error messages, but my quantized ONNX model has this output:
![image](https://github.com/amd/RyzenAI-SW/assets/109336572/f1668b9c-ad54-4e85-8e5a-e3de92fe9de2)

And the functioning example in tutorial\yolov8_e2e\implement has this output:
![image](https://github.com/amd/RyzenAI-SW/assets/109336572/0a5c43e9-9ea6-4c1a-bd4c-25a2123d0798)

Unfortunately, the missing output (name: 3604 tensor: float32[1,84,8400]) is the one with the detections. I have tried many solutions, but have wasted a lot of time. Does anyone have an idea how to fix the issue?

Any ideas are welcome :)

**Comment by xXxTheUserxXx** [comment]
I have a warning that I received, and I have highlighted what I find interesting in bold. Perhaps someone has an idea if this could be the problem and in which direction I could look to solve my described problem.

____________________________________________________________________________________________________________________________


/opt/vitis_ai/conda/envs/vitis-ai-pytorch/lib/python3.8/site-packages/torch/onnx/utils.py:2040: UserWarning: No names were found for specified dynamic axes of provided input.Automatically generated names will be applied to each dynamic axes of input DetectionModel::input_0
  warnings.warn(
/opt/vitis_ai/conda/envs/vitis-ai-pytorch/lib/python3.8/site-packages/pytorch_nndct/nn/modules/prim_ops.py:116: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if not (list(self.node.out_tensors[0].shape[1:]) == list(input.size())[1:]):
/opt/vitis_ai/conda/envs/vitis-ai-pytorch/lib/python3.8/site-packages/pytorch_nndct/quantization/quantizerimpl.py:17: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if inf.sum() > 0 or nan.sum() > 0:
/opt/vitis_ai/conda/envs/vitis-ai-pytorch/lib/python3.8/site-packages/pytorch_nndct/nn/modules/fix_ops.py:67: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if (isinstance(tensor, torch.Tensor) and
/opt/vitis_ai/conda/envs/vitis-ai-pytorch/lib/python3.8/site-packages/torch/onnx/_internal/jit_utils.py:258: UserWarning: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function. (Triggered internally at ../torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1884.)
  _C._jit_pass_onnx_node_shape_type_inference(node, params_dict, opset_version)
/opt/vitis_ai/conda/envs/vitis-ai-pytorch/lib/python3.8/site-packages/torch/onnx/utils.py:687: UserWarning: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function. (Triggered internally at ../torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1884.)
  _C._jit_pass_onnx_graph_shape_type_inference(
/opt/vitis_ai/conda/envs/vitis-ai-pytorch/lib/python3.8/site-packages/torch/onnx/utils.py:1178: **UserWarning: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function. (Triggered internally at ../torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1884.)
  _C._jit_pass_onnx_graph_shape_type_inference(**

[VAIQ_NOTE]: DetectionModel_int.onnx is generated.(/workspace/RyzenAI-SW/runs/detect/train/weights/nndct_quant/DetectionModel_int.onnx)

[VAIQ_NOTE]: =>Converting to xmodel ...

[VAIQ_NOTE]: =>Dumping 'DetectionModel_0'' checking data...

[VAIQ_NOTE]: =>Finish dumping data.(/workspace/RyzenAI-SW/runs/detect/train/weights/nndct_quant/deploy_check_data_int/DetectionModel_0)

[VAIQ_NOTE]: =>Successfully convert 'DetectionModel_0' to xmodel.(/workspace/RyzenAI-SW/runs/detect/train/weights/nndct_quant/DetectionModel_0_int.xmodel)

**Comment by xXxTheUserxXx** [comment]
I'm still unable to quantize my own YOLOv8m model successfully. Although my quantized model appears to be good, it lacks the final layers present in the example provided in RyzenAI-SW\tutorial\yolov8_e2e\DetectionModel_int.onnx.  

This part is missing in my quantized model:

![image](https://github.com/amd/RyzenAI-SW/assets/109336572/db9a703c-216f-4349-98a9-88a0951db7a0)

Unfortunately, I need to get this to work. What kind of error could lead to this? I am grateful for any ideas.

**Comment by uday610** [comment]
Hello @xXxTheUserxXx , 

We are checking the warning message if that provides any hint, we will let you know... 

But the yolov8m model which we provided in the tutorial does not show this issue. So I wonder what is different in your yolov8m model or how you prepare it? If there is any way we can try the same model you are trying? 

Thanks 
Uday

**Comment by xXxTheUserxXx** [comment]
Hello @uday610,

I encountered the same issue when following [4 Quantization](https://github.com/amd/RyzenAI-SW/blob/main/tutorial/yolov8_e2e/README.md#4-quantization). I believe you should also be able to replicate the problem.

What I do (I didn't mention some steps to simplify reading):
1) Start an AWS EC2 instance g5.2xlarge -> This instance features a NVIDIA A10G graphics card.
2) Download and unzip ipu-rel-3.5.0-246.tar.gz.
3) Build a Docker image using "docker_build.sh -t gpu -f pytorch".

The following is a bit different from what's explained in the example:

4)	`docker run -it -v ./:/workspace --gpus all --shm-size 16G xilinx/vitis-ai-ipu-pytorch-gpu:3.5.0.246`
5)	`sudo nano -c run_ptq.sh`
6)	Add `batch=1 dump_onnx=True` in line 29 of run_ptq.sh:
```
       echo "[Test mode] testing"
       CUDA_VISIBLE_DEVICES=${GPU_ID} yolo detect val data="datasets/coco.yaml" model=${WEIGHTS} \
       nndct_quant=True quant_mode=test batch=1 dump_onnx=True --nndct_convert_sigmoid_to_hsigmoid --nndct_convert_silu_to_hswish
```

7)	`bash run_ptq.sh`
______________________________________________________________________________________________________________


I recommend the following steps to limit terminal outputs for easier spotting of the warnings I posted:

1) `sudo nano -c /workspace/tutorial/yolov8_e2e/code/ultralytics/yolo/engine/validator.py`
2) `verbose=False` in line 283 and 287.
______________________________________________________________________________________________________________

Could you please export your own `DetectionModel_int.onnx` and check if your model also has a missing output compared to the model in `/RyzenAI-SW/tutorial/yolov8_e2e/implement/DetectionModel_int.onnx`?

**Comment by xXxTheUserxXx** [comment]
Hi @uday610,

I've attached [my_DetectionModel_int.zip](https://github.com/amd/RyzenAI-SW/files/14295630/my_DetectionModel_int.zip), which is the YOLOv8m model I quantized as described in the previous comment. Unfortunately, I'm unable to share the model we trained, but they both have the same issue.

Maybe this will help in figuring something out. I also think I am not the only one having this issue. So maybe it is also in your interest to improve RyzenAI.

Thank you for your help!

**Comment by kewang-xlnx** [comment]
Hi @xXxTheUserxXx :

The output (name: 3604 tensor: float32[1,84,8400]) is generated from the post_process of DetectionModel. In fp32 onnx graph, the post_process part is the subgraph you have mentioned above.

And in quantized model, we have excluded this part in DetectionModel, in line 63 of file code/ultralytics/nn/tasks.py. And we added post_process after the quantized model generated, in line 233 of file code/ultralytics/yolo/engine/validator.py. The post_process will not be deployed in IPU device.

If we keep post_process in quantized model, the accuracy of yolov8m will be very low. And if you want to include it in the quantized mode, you could try to modify the script and requantized it.

**Comment by xXxTheUserxXx** [comment]
Hello, everyone! Thanks to your help, I managed to find a solution.
